### PR TITLE
FSDK-63: Typescript Utils

### PR DESCRIPTION
**Pull request by mahboubii**

### CODE REVIEW

File: src/utils/errors.ts [hunk 1]
```diff
@@ -1,12 +1,28 @@
...
+  | 'delete-reaction'
+  | 'add-child-reaction'
+  | 'delete-child-reaction';
 
-export const handleError = (error: Error, type: string, detail: ErrorDetail) => {
+export type ErrorDetail = { kind?: string };
+
+export type ErrorHandler = (error: Error, type: NetworkRequestTypes, details: ErrorDetail) => void;
+
+export const handleError = (error: Error, type: NetworkRequestTypes, detail: ErrorDetail) => {
```
Comment (arnautov-anton) [inline src/utils/errors.ts:21]:
```suggestion
export const handleError: ErrorHandler = (error, type, detail) => {
```

File: src/utils/index.ts [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+  const now = tDateTimeParser();
+  return time.from(now);
+}
+
+type ErrorUser = { error: string };
+function isErrorUser(user: unknown | ErrorUser): user is ErrorUser {
+  return !!user && typeof (user as ErrorUser).error === 'string';
+}
+
+export function userOrDefault<T>(user?: T | string | { error: string } | null) {
```
Comment (arnautov-anton) [inline src/utils/index.ts:35]:
```suggestion
export function userOrDefault<T extends UserAPIResponse>(user?: T | string | { error: string } | null) {
```

File: src/utils/textRenderer.tsx [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+  const renderedText = text
+    .split(/\r\n|\r|\n/) // first break on line
+    .map((line, i) =>
+      line
+        .split(' ') // break for each word
+        .map((word, j) => renderWord(word, `item-${i}-${j}`, parentClass, onClickMention, onClickHashtag))
+        .reduce<WordArray>((acc, elem) => (acc.length ? [acc, ' ', elem] : [elem]), []),
+    )
+    .reduce<WordArrayArray>((acc, elem) => (acc.length ? [acc, '\n', elem] : [elem]), []);
```
Comment (arnautov-anton) [inline src/utils/textRenderer.tsx:109]:
```suggestion
  const reducer = (separator?: string = ' ') => (acc, elem) => (acc.length ? [acc, separator, elem] : [elem])

  const renderedText = text
    .split(/\r\n|\r|\n/) // first break on line
    .map((line, i) =>
      line
        .split(' ') // break for each word
        .map((word, j) => renderWord(word, `item-${i}-${j}`, parentClass, onClickMention, onClickHashtag))
        .reduce<WordArray>(reducer(), []),
    )
    .reduce<WordArrayArray>(reducer('\n'), []);
```

File: src/utils/index.ts (hunk 2 above)
Comment (mahboubii) [reply src/utils/index.ts:35]: `UserAPIResponse` also needs a generic to work correctly, let's revisit this later if needed

File: src/utils/errors.ts (hunk 1 above)
Comment (mahboubii) [reply src/utils/errors.ts:21]: fixed 👍

File: src/utils/textRenderer.tsx [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+  const [link] = linkify.find(word);
+  if (!link) return word;
+
+  const { type, value, href } = link;
+
+  if (onClickMention && type === 'mention') {
+    return (
+      <CustomAnchor
+        key={key}
+        type="mention"
```
Comment (mahboubii) [inline src/utils/textRenderer.tsx:74]:
```suggestion
        type={type}
```

File: src/utils/textRenderer.tsx [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        parentClass={parentClass}
+      />
+    );
+  }
+
+  if (onClickHashtag && type === 'hashtag') {
+    return (
+      <CustomAnchor
+        key={key}
+        type="hashtag"
```
Comment (mahboubii) [inline src/utils/textRenderer.tsx:87]:
```suggestion
        type={type}
```

**Pull request merged**